File: simplified-books-controller/src/main/java/org/nypl/simplified/books/controller/ProfileSelectionTask.java
package org.nypl.simplified.books.controller;

import com.io7m.jfunctional.Unit;
import com.io7m.jnull.NullCheck;

import org.nypl.simplified.observable.ObservableType;
import org.nypl.simplified.profiles.api.ProfileAnonymousEnabledException;
import org.nypl.simplified.profiles.api.ProfileEvent;
import org.nypl.simplified.profiles.api.ProfileID;
import org.nypl.simplified.profiles.api.ProfileNonexistentException;
import org.nypl.simplified.profiles.api.ProfileSelected;
import org.nypl.simplified.profiles.api.ProfilesDatabaseType;

import java.util.concurrent.Callable;

import static org.nypl.simplified.profiles.api.ProfilesDatabaseType.AnonymousProfileEnabled.ANONYMOUS_PROFILE_ENABLED;

final class ProfileSelectionTask implements Callable<Unit> {

  private final ProfilesDatabaseType profiles;
  private final ProfileID profile_id;
  private final ObservableType<ProfileEvent> profile_events;

  ProfileSelectionTask(
      final ProfilesDatabaseType in_profiles,
      final ObservableType<ProfileEvent> in_events,
      final ProfileID in_id) {

    this.profiles =
        NullCheck.notNull(in_profiles, "Profiles");
    this.profile_events =
        NullCheck.notNull(in_events, "Events");
    this.profile_id =
        NullCheck.notNull(in_id, "ID");
  }

  @Override
  public Unit call() throws ProfileNonexistentException, ProfileAnonymousEnabledException {
    if (this.profiles.anonymousProfileEnabled() == ANONYMOUS_PROFILE_ENABLED) {
      this.profile_events.send(ProfileSelected.of());
      return Unit.unit();
    }

    this.profiles.setProfileCurrent(this.profile_id);
    this.profile_events.send(ProfileSelected.of());
    return Unit.unit();
  }
}

File: simplified-books-controller/src/main/java/org/nypl/simplified/books/controller/ProfileSelectionTask.kt
package org.nypl.simplified.books.controller

import io.reactivex.subjects.Subject
import org.joda.time.LocalDateTime
import org.nypl.simplified.analytics.api.AnalyticsEvent
import org.nypl.simplified.analytics.api.AnalyticsType
import org.nypl.simplified.books.book_registry.BookRegistryType
import org.nypl.simplified.profiles.api.ProfileAnonymousEnabledException
import org.nypl.simplified.profiles.api.ProfileEvent
import org.nypl.simplified.profiles.api.ProfileID
import org.nypl.simplified.profiles.api.ProfileNonexistentException
import org.nypl.simplified.profiles.api.ProfileSelection
import org.nypl.simplified.profiles.api.ProfileSelection.ProfileSelectionInProgress
import org.nypl.simplified.profiles.api.ProfilesDatabaseType
import org.nypl.simplified.profiles.api.ProfilesDatabaseType.AnonymousProfileEnabled.ANONYMOUS_PROFILE_ENABLED
import org.slf4j.LoggerFactory
import java.util.concurrent.Callable

class ProfileSelectionTask(
  private val analytics: AnalyticsType,
  private val bookRegistry: BookRegistryType,
  private val events: Subject<ProfileEvent>,
  private val id: ProfileID,
  private val profiles: ProfilesDatabaseType
) : Callable<Unit> {

  private val logger = LoggerFactory.getLogger(ProfileSelectionTask::class.java)

  @Throws(ProfileNonexistentException::class, ProfileAnonymousEnabledException::class)
  override fun call() {
    try {
      this.logger.debug("[{}]: profile selection in progress", this.id.uuid)
      if (this.profiles.anonymousProfileEnabled() === ANONYMOUS_PROFILE_ENABLED) {
        this.events.onNext(ProfileSelectionInProgress(this.id))
        return this.loadData()
      }

      this.profiles.setProfileCurrent(this.id)
      this.events.onNext(ProfileSelectionInProgress(this.id))
      return this.loadData()
    } finally {
      this.logger.debug("[{}]: profile selection completed", this.id.uuid)
      this.events.onNext(ProfileSelection.ProfileSelectionCompleted(this.id))
      this.publishAnalyticsEvent()
    }
  }

  private fun publishAnalyticsEvent() {
    val profile = this.profiles.currentProfileUnsafe()
    this.analytics.publishEvent(
      AnalyticsEvent.ProfileLoggedIn(
        timestamp = LocalDateTime.now(),
        profileUUID = profile.id.uuid,
        displayName = profile.displayName,
        birthDate = profile.preferences().dateOfBirth?.show(),
        attributes = profile.attributes().attributes
      )
    )
  }

  private fun loadData() {
    return ProfileDataLoadTask(
      profile = this.profiles.currentProfileUnsafe(),
      bookRegistry = this.bookRegistry
    )
      .run()
  }
}
